Midichlorians/Server/test.cc

```diff
@@ -0,0 +1,313 @@
+    }
+  }
+
+  void operator()(const iOSGenericEvent& event) {
+    std::string params = "source=" + event.source + " event=" + event.event;
+    for (const auto& f : event.fields) {
+      params += ' ';
+      params += f.first + '=' + f.second;
+    }
+    AppendToLastEvent(Printf("[iOSGenericEvent user_ms=%lld ", event.user_ms.count()) + params + ']');
+  }
+
+  void operator()(const iOSBaseEvent&) {}
+
+  // Web events.
+  std::string ExtractWebBaseEventParams(const WebBaseEvent& e) {
+    std::string result;
+    result += "customer_id=" + e.customer_id + ' ';
+    result += "user_ms=" + ToString(e.user_ms.count()) + ' ';
+    result += "client_id=" + e.client_id + ' ';
+    result += "referer_host=" + e.referer_host + ' ';
+    result += "referer_path=" + e.referer_path + ' ';
+    bool first = true;
+    for (const auto& cit : e.referer_querystring) {
+      if (first) {
+        first = false;
+      } else {
+        result += ' ';
+      }
+      result += cit.first + '=' + cit.second;
+    }
+    return result;
+  }
+
+  void operator()(const WebForegroundEvent& e) {
+    AppendToLastEvent("[WebForegroundEvent " + ExtractWebBaseEventParams(e) + ']');
+  }
+
+  void operator()(const WebBackgroundEvent& e) {
+    AppendToLastEvent("[WebBackgroundEvent " + ExtractWebBaseEventParams(e) + ']');
+  }
+
+  void operator()(const WebEnterEvent& e) {
+    AppendToLastEvent("[WebEnterEvent " + ExtractWebBaseEventParams(e) + ']');
+  }
+
+  void operator()(const WebExitEvent& e) {
+    AppendToLastEvent("[WebExitEvent " + ExtractWebBaseEventParams(e) + ']');
+  }
+
+  void operator()(const WebGenericEvent& e) {
+    AppendToLastEvent("[WebGenericEvent " + e.event_action + ':' + e.event_category + ' ' +
+                      ExtractWebBaseEventParams(e) + ']');
+  }
+
+  void AppendToLastEvent(const std::string& what) {
+    assert(events.size());
+    std::string& last = events.back();
+    last += what;
+  }
+
+  std::atomic_size_t total_count;
+  std::vector<std::string> events;
+  std::vector<std::string> errors;
+};
+
+TEST(MidichloriansServer, iOSEventsFromCPP) {
+  using namespace current::midichlorians::server;
+  using namespace current::midichlorians::ios;
+
+  current::time::SetNow(std::chrono::microseconds(0));
+  GenericConsumer consumer;
+  MidichloriansHTTPServer<GenericConsumer> server(
+      FLAGS_server_test_port, consumer, std::chrono::milliseconds(100), "/log", "OK\n");
+  const std::string server_url = Printf("http://localhost:%d/log", FLAGS_server_test_port);
+
+  current::time::SetNow(std::chrono::microseconds(12000));
+  iOSAppLaunchEvent launch_event;
+  launch_event.user_ms = std::chrono::milliseconds(1);
+  T_IOS_VARIANT event1(std::move(launch_event));
+  const auto response1 = HTTP(POST(server_url, JSON(event1) + "\nblah"));
+  EXPECT_EQ(200, static_cast<int>(response1.code));
+  EXPECT_EQ("OK\n", response1.body);
+
+  current::time::SetNow(std::chrono::microseconds(112000));
+  while (consumer.total_count < 3u) {
```

Comment:
It's not obvious why `3u`. Perhaps split this test into several, to keep them unit?